fleekxyz/non-fungible-apps
# feat: verifications and mongo write operations

Pr 253 is ready for review. This is a big pr because of the changes it brings to subgraph, contract, and serverless. There are some breaking changes for the UI so I will share some notes on what to look out for when the PR is merged.

Contract

- The metadataUpdate event with two string values now has four. It's going to emit ipfsHash, domain, gitRepository, and commit hash for each build. Subgraph's been adjusted to handle it.
- The setTokenBuild function now takes two additional input values: ipfsHash and domain

Subgraph

- We have a new builds field in the token entity that has these values: 
    - gitRepository
    - commitHash
    - ipfsHash
    - domain

- The gitRepository and commitHash fields are removed from the token entity. The NewMint event and entity both have ipfsHash from now on.
Serverless

- Both build and mint handlers are completed (at least from the verifications and database aspects).

    -> Mint: The handler checks and decodes the block log data that are sent to it from the Alchemy Notify service. Then queries the database with Prisma to check the builds collection for any build record with the same build information as passed in the mint parameters. If there is anything, it will mark the token as verified in the database and then make a contract call to the NFA contract. I was planning on leaving this part as a to-do for the future as discussed in the previous sync but to make the flow more complete continued working on it and now the transaction is signed and sent to the contract. I just need to test it on-chain. Mining some goerli eth for a new deployment. Finally, the mint handler adds the token to the tokens collection and returns some data on the request back to the entity that initiated it.

    -> Build: Like mint, this flow first checks the request body but doesn't decode it (not necessary since it's not encoded). Then structures and adds data to the database and looks for a matching token in the tokens collection. If it finds anything, it will mark it as verified and make a call to the smart contract to update the status there. The verified status in the Mongodb will be updated, too.

How to test it locally

I have written a readme but it's not very complete so first make sure you have MongoDB installed. then you need to run a replica set of Mongodb and initiate it from the Mongo shell. you can populate it with some data but I don't think it's necessary since the Prisma schema is already included in the code. So simply you need to update the .env file and change the database_url value with the API URL and private key.

As soon as you do these things, you can start a serverless offline instance with yarn start and in another terminal send requests.

These two are sample requests for build and mint, but they probably will not work with the latest commit (the mint params need to be updated with the new contract deployment).

The mint request is a simulation of what is sent from Alchemy notify and the build request is an example of what we should expect from fleek workflows that manage builds.

```
Build request

# curl --header "Content-Type: application/json" --request POST --data '{"githubRepository":"repo", "commitHash":"hash","tokenId":14,"ipfsHash":"a1a1a1a1", "domain":"https://fleek.nfa"}' http://localhost:3000/prd/build

Mint request

curl --header "Content-Type: application/json" --request POST --data '{"webhookId":"wh_gsqn2uadyt8uea6a","id":"whevt_ydckqph5c0fp9mkk","createdAt":"2023-04-27T11:07:15.464Z","type":"GRAPHQL","event":{"data":{"block":{"hash":"0xc76ab7a9700dc62211db3221c52416517c5d637d9d6f06972a2629111751d52d","number":8900182,"timestamp":1682593632,"logs":[{"data":"0x","topics":["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef","0x0000000000000000000000000000000000000000000000000000000000000000","0x000000000000000000000000d4997d0facc83231b9f26a8b2155b4869e99946f","0x000000000000000000000000000000000000000000000000000000000000000e"],"index":77,"account":{"address":"0x8795608346eb475e42e69f1281008aeaa522479d"},"transaction":{"hash":"0x9fc7bdf15e6c8c854512d853a0850476e1a8833f4fb7d617f0e456e38fb926e0","nonce":13,"index":56,"from":{"address":"0xd4997d0facc83231b9f26a8b2155b4869e99946f"},"to":{"address":"0x8795608346eb475e42e69f1281008aeaa522479d"},"value":"0x0","gasPrice":"0x1d2a7b6350","maxFeePerGas":"0x2714326c4f","maxPriorityFeePerGas":"0x59682f00","gas":477643,"status":1,"gasUsed":314007,"cumulativeGasUsed":7605215,"effectiveGasPrice":"0x1d2a7b6350","createdContract":null}},{"data":"0x0000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000001c000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000240000000000000000000000000000000000000000000000000000000000000028000000000000000000000000000000000000000000000000000000000000002c000000000000000000000000000000000000000000000000000000000000030390000000000000000000000000000000000000000000000000000000000000001000000000000000000000000d4997d0facc83231b9f26a8b2155b4869e99946f0000000000000000000000000000000000000000000000000000000000000005666c65656b00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000046465736300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000375726c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000465656e73000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004686173680000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000047265706f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000046c6f676f00000000000000000000000000000000000000000000000000000000","topics":["0xba754491b4e452f016c3b656dfb44a5933d928f4dabae3cf2d65f1a27e4b8072","0x000000000000000000000000000000000000000000000000000000000000000e","0x000000000000000000000000d4997d0facc83231b9f26a8b2155b4869e99946f","0x000000000000000000000000d4997d0facc83231b9f26a8b2155b4869e99946f"],"index":78,"account":{"address":"0x8795608346eb475e42e69f1281008aeaa522479d"},"transaction":{"hash":"0x9fc7bdf15e6c8c854512d853a0850476e1a8833f4fb7d617f0e456e38fb926e0","nonce":13,"index":56,"from":{"address":"0xd4997d0facc83231b9f26a8b2155b4869e99946f"},"to":{"address":"0x8795608346eb475e42e69f1281008aeaa522479d"},"value":"0x0","gasPrice":"0x1d2a7b6350","maxFeePerGas":"0x2714326c4f","maxPriorityFeePerGas":"0x59682f00","gas":477643,"status":1,"gasUsed":314007,"cumulativeGasUsed":7605215,"effectiveGasPrice":"0x1d2a7b6350","createdContract":null}}]}},"sequenceNumber":"10000000004053708000"}}' http://localhost:3000/prd/mint

```
